messybrainz/create_recording_clusters.py

```diff
@@ -0,0 +1,208 @@
+    """Checks if recording cluster is already present in the recording_redirect table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id: the gid which represents the cluster.
+        mbid: mbid for the cluster.
+
+    Returns:
+        True if cluster already present in recording_redirect table otherwise 
+        False is returned.
+    """
+
+    query = text("""SELECT recording_cluster_id, recording_mbid
+                      FROM recording_redirect
+                     WHERE recording_cluster_id = :recording_cluster_id
+                       AND recording_mbid = :recording_mbid
+    """)
+    result = connection.execute(query, {
+        "recording_cluster_id": cluster_id,
+        "recording_mbid": mbid,
+    })
+    if result.rowcount:
+        return True
+    return False
+
+
+def is_recording_cluster_present_in_recording_cluster(connection, cluster_id, gid):
+    """Checks if recording cluster is present in the recording_cluster table.
+
+    Args:
+        connection: the sqlalchemy db connection to be used to execute queries
+        cluster_id: the gid which represents the cluster.
+        gid: gid for a recording.
+
+    Returns:
+        True if cluster present in recording_cluster table otherwise 
+        False is returned.
+    """
+
+    query = text("""SELECT cluster_id, recording_gid
+                      FROM recording_cluster
+                     WHERE cluster_id = :cluster_id
+                       AND recording_gid = :recording_gid
+    """)
+    result = connection.execute(query, {
+        "cluster_id": cluster_id,
+        "recording_gid": gid,
+    })
+    if result.rowcount:
+        return True
+    return False
+
+
+def truncate_tables(connection):
+    """Truncates recording_cluster and recording_redirect tables."""
+
+    query = text("TRUNCATE TABLE recording_cluster")
+    connection.execute(query)
+    query = text("TRUNCATE TABLE recording_redirect")
+    connection.execute(query)
+
+
+def msid_processed(connection):
+    """Returns the number of MSIDs processed by this script."""
+
+    query = text("SELECT COUNT(*) FROM recording_cluster")
+    result = connection.execute(query)
+    num_msid = result.fetchone()
+    return num_msid[0]
+
+
+def create_recording_clusters(reset=False):
+    """Creates clusters for recording mbids present in the recording_json table.
+
+    Args:
+        reset (boolean): should recording_cluster and recording_redirect tables be
+                         truncated before creating clusters or not. 
+    Returns:
+        clusters_modified (int): number of clusters modified by the script.
+        clusters_add_to_redirect (int): number of clusters added to redirect table.
+        num_msid_processed (int): number of MSIDs processed by the script.
+    """
+
+    clusters_modified = 0
+    clusters_add_to_redirect = 0
+    num_msid_processed = 0
+    with db.engine.begin() as connection:
+        if reset:
```

Comment:
This reset functionality makes me nervous -- for putting this into production is could lead to disaster -- but for development this is fine for now. Later, once we're past the experimentation stage of this project we should consider removing these features, or at least adding some safeguards so that the tables can't be truncated in a production environment.